File: 01-Beginner-Client-libraries/08-Implementing-custom-interfaces/ros2_ws/src/more_interfaces/scripts/subscribe_contact.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from more_interfaces.msg import Contact


class ContactSubscriber(Node):

    def __init__(self):
        super().__init__('address_book_subscriber_py')
        self.subscription_ = self.create_subscription(
            Contact,
            'address_book',
            self.listener_callback,
            10)
        self.subscription_

    def listener_callback(self, msg):
        self.get_logger().info(
            f'Received: {msg.first_name} {msg.last_name} | '
            f'phone: {msg.phone_number} | type: {msg.phone_type}')


def main():
    rclpy.init()
    node = ContactSubscriber()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: 01-Beginner-Client-libraries/08-Implementing-custom-interfaces/ros2_ws/src/more_interfaces/src/subscribe_contact.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "more_interfaces/msg/contact.hpp"


class ContactSubscriber : public rclcpp::Node {
    private:
        rclcpp::Subscription<more_interfaces::msg::Contact>::SharedPtr contact_subscription_;
    public:
        ContactSubscriber() : Node("contact_subscriber") {
            contact_subscription_ = this->create_subscription<more_interfaces::msg::Contact>("contact", 10, 
            [this] (const more_interfaces::msg::Contact::SharedPtr msg) -> void {
                RCLCPP_INFO(this->get_logger(),
                    "Received: %s %s | phone: %s | type: %d",
                    msg->first_name.c_str(),
                    msg->last_name.c_str(),
                    msg->phone_number.c_str(),
                    msg->phone_type);
            });
        }
};


int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ContactSubscriber>());
    rclcpp::shutdown();
    return 0;
}